wallet-new/server/Cardano/Wallet/API/V1/LegacyHandlers/Addresses.hs

```diff
@@ -25,21 +30,32 @@ handlers =  listAddresses
        :<|> newAddress
        :<|> verifyAddress
 
+-- | This is quite slow. What happens when we have 50k addresses?
+-- TODO(ks): One idea I have is to persist the length of the
+-- addresses and send that to `respondWith`,
+-- while we could lazily fetch the data (and have better performance)
+-- we need to show what the current page contains?
+-- In other words, we wouldn't send everything to `respondWith`.
```

Comment:
this is where I pipe up and say "let's please use a SQL database for this sort of thing" :smile: SQLite or bundling Postgres would be *awesome* and these problems entirely go away.

Reply:
Thanks, I'll check it out.

@parsonsmatt I agree.

Reply:
@adinapoli-iohk I checked your example out. It does seem relatively fast (it's around 5 seconds on my machine), but I think we still aren't addressing the biggest issue here, which should be the reading/deserialization? Once we have that in the picture, the performance should degrade even more rapidly. 
Paging or SQL, as Matt proposed, seem like the options we can think about for the future. Or caching :man_shrugging:

Reply:
> SQLite or bundling Postgres would be awesome and these problems entirely go away.

Yes, perhaps SQL might be a good/better fit (and I'm not opposed to review our persistence options) but not after we have @dcoutts back. Generally speaking, using a relational database might need innocent enough, but I would humbly propose to not disregard the advantages of using `acid-state` here, either. In particular, `acid-state` gives you acid guarantees on your data structures (almost) for free without the need to even worry about SQL concepts like transactions & rollbacks. Query a data store is basically a pure operation (in our codebase a `Query` is essentially a synonym for `ReaderT`) and you get a clean separation between read & writes in your persistence logic, which is not something obvious. You get data migration via `safecopy` and you don't need to tap into external solutions (and consequently deal with running the migration as part of shipping the installer, which complicates bundling this to customer even further). Last but not least, there is the hidden cost of actually writing migrations for `acid-state` to a relational DB and ship that in the field. Not saying `acid-state` is perfect (far from it) but having used it commercially in the past, I can totally stand for it as a viable solution 😉 Of course it has its shortcomings and I'm happy to review the situation, but not before Duncan is back, as I have said 😁 

> but I think we still aren't addressing the biggest issue here, which should be the reading/deserialization?

Uhm, I'm not sure I'm getting you here -- what do you mean exactly? `acid-state` is an in-memory DB so obviously the assumption is that you can (basically) fit most of your data in primary memory, so once is in RAM it's relatively easy (and fast) to access. If by deserialisation you mean the code of _deserialising_ the addresses, I think that was a flaw of the previous design: we should never have stored `CId Addr` into the `WalletStorage` to begin with. In fact, see https://github.com/input-output-hk/cardano-sl/pull/2675 😉

Reply:
I'm happy to punt on the exact persistence layer :) Perhaps we can even figure out a way to use `acid-state` in a more relational manner!

Reply:
> acid-state is an in-memory DB

We have to read/write from and to the DB, otherwise, it's not a DB. Also, I think we are using it in "disk mode" in production (https://hackage.haskell.org/package/acid-state-0.14.3/docs/Data-Acid-Local.html).

Reply:
> We have to read/write from and to the DB, otherwise, it's not a DB

I'm not an expert, but my intuition is that it works a bit like Redis and other similar DBs: you keep in memory a transaction log of your operations (this is very fast as it's RAM I/O) and periodically you dump this to this (this is called "checkpointing").